addons/stock/models/stock_quant.py

```diff
@@ -277,19 +277,25 @@ def _get_available_quantity(self, product_id, location_id, lot_id=None, package_
 
     @api.onchange('location_id', 'product_id', 'lot_id', 'package_id', 'owner_id')
     def _onchange_location_or_product_id(self):
```

Comment:
It's more readable this way (but still doesn't work)

``` python
vals = {}
# A lot_id is set: perform a sanity check and prefill the inventory quantity to 1
# if the tracking is of serial type.
if self.lot_id:
    if self.tracking == 'none' or self.product_id != self.lot_id.product_id:
        vals['lot_id'] = None
    elif self.tracking == 'serial':
        vals['inventory_quantity'] = 1

# A product and a location is set: make sure the reserved quantity is up to date.
if self.product_id and self.location_id:
    quants = self._gather(self.product_id, self.location_id, lot_id=self.lot_id, package_id=self.package_id, owner_id=self.owner_id, strict=True)
    quantity_on_hand = sum(quants.mapped('quantity'))
    reserved_quantity = sum(quants.mapped('reserved_quantity'))
    vals['quantity'] = quantity_on_hand
    vals['reserved_quantity'] = reserved_quantity

if vals:
    self.update(values_to_update)
```

Maybe perform all the sanity check at the start and all the inventoryq ty logic at the end?

Reply:
I must keep the condition `if self.inventory_quantity == self.quantity` to avoid erase the `inventory_quantity` if user modified it.
So, I have this:
``` python3
vals = {}
# A lot_id is set: perform a sanity check
if self.lot_id:
    if self.tracking == 'none' or self.product_id != self.lot_id.product_id:
        vals['lot_id'] = None

# Product and location are set: make sure the reserved quantity is up to date.
if self.product_id and self.location_id:
    quants = self._gather(self.product_id, self.location_id, lot_id=self.lot_id, package_id=self.package_id, owner_id=self.owner_id, strict=True)
    vals['reserved_quantity'] = sum(quants.mapped('reserved_quantity'))

    # In case of SN tracked product, we force `product_qty` to 1 because
    # it's the only relevant value aside 0 for this kind of product.
    if self.lot_id and self.product_id.tracking == 'serial':
        vals['quantity'] = vals['inventory_quantity'] = 1
    # Otherwise, we update quantity only if user didn't modify it manually.
    elif self.inventory_quantity == self.quantity:
        quantity_on_hand = sum(quants.mapped('quantity'))
        vals['quantity'] = vals['inventory_quantity'] = quantity_on_hand
if vals:
    self.update(vals)
```

Reply:
``` python
vals = {}
# A lot_id is set: perform a sanity check.
if self.lot_id:
    if self.tracking == 'none' or self.product_id != self.lot_id.product_id:
        vals['lot_id'] = None

# Product and location are set: make sure the reserved quantity is up to date.
if self.product_id and self.location_id:
    quants = self._gather(self.product_id, self.location_id, lot_id=self.lot_id, package_id=self.package_id, owner_id=self.owner_id, strict=True)
    vals['reserved_quantity'] = sum(quants.mapped('reserved_quantity'))
    # Note: we update `quantity` even if it doesn't make sense. This is used to compare
    # with the inventory quantity to know if the user updates the quantity manually. It relies
    # on the fact the field is hidden in the view and not sent to the server.
    vals['quantity'] = sum(quants.mapped('quantity'))

    # If the user didn't change the inventory quantity, update it to the new quantity.
    if float_compare(vals['quantity'], self.inventory_quantity, precision_rounding=self.uom_id.rounding):
        vals['quantity'] = vals['inventory_quantity'] 
if vals:
    self.update(vals)
```

Reply:
@svs-odoo 
``` python
    @api.onchange('location_id', 'product_id', 'lot_id', 'package_id', 'owner_id')
    def _onchange_location_or_product_id(self):
        vals = {}

        # Keep `inventory_quantity` only if the user manually updated it. The user should not
        # interact with `self.quantity`.
        update_inventory_quantity = float_compare(self.inventory_quantity, self.quantity, precision_rounding=self.uom_id.rounding) == 0
        reset_lot_id = False

        # Sanity check if a lot has been set.
        if self.lot_id:
            if self.tracking == 'none' or self.product_id != self.lot_id.product_id:
                reset_lot_id = True

        # Once the new line is complete, fetch the new theoretical values.
        if self.product_id and self.location_id:
            quants = self._gather(self.product_id, self.location_id, lot_id=self.lot_id, package_id=self.package_id, owner_id=self.owner_id, strict=True)
            reserved_quantity = sum(quants.mapped('reserved_quantity'))
            quantity = sum(quants.mapped('quantity'))

            vals['reserved_quantity'] = reserved_quantity
            vals['quantity'] = quantity
            if update_inventory_quantity:
                vals['inventory_quantity'] = quantity

            if reset_lot_id:
                vals['lot_id'] = None

            # Special case: directly set the quantity to one for serial numbers.
            if self.lot_id and self.product_id.tracking == 'serial' and not reset_lot_id and update_inventory_quantity:
                vals['inventory_quantity'] = 1

        if vals:
            self.update(vals)
```